File: Chores/engineering/merge.py
import requests
import re
import os
from datetime import datetime
import yaml

with open('Chores/engineering/data/sgmodules.yaml', 'r') as f:
    sgmodule_info = yaml.safe_load(f)

sections = {
    "URL Rewrite": [],
    "Map Local": [],
    "Script": [],
    "MITM": [],
    "Rule": []
}

section_pattern = re.compile(r'\[(.*?)\]\s*\n(.*?)(?=\n\[|$)', re.DOTALL)
hostname_pattern = re.compile(r'hostname\s*=\s*(.*)', re.IGNORECASE)
headers = []

max_divider_length = 30  # 最大分隔符长度（不包括header的长度）

for info in sgmodule_info:
    try:
        response = requests.get(info['url'])
        response.raise_for_status()
        
        content = response.text
        print(f"[Debug] Parsing content from {info['header']}")
        matches = section_pattern.findall(content)
        
        headers.append(info['header'])
        
        # 计算动态分隔线
        left_dash_count = (max_divider_length - len(info['header'])) // 2
        right_dash_count = max_divider_length - len(info['header']) - left_dash_count
        divider = f"# {'-' * left_dash_count} {info['header']} {'-' * right_dash_count}"
        
        for section, text in matches:
            if section in sections and section != "MITM":
                if section == "Rule":
                    cleaned_text = re.sub(r",\s*(REJECT|DIRECT)", "", text).strip()
                    sections["Rule"].append(cleaned_text)
                    print(f"[Debug] Added cleaned Rule content from {info['header']}: {cleaned_text}")
                else:
                    # 删除内容中的注释行和空行，但保留 divider
                    cleaned_lines = "\n".join(
                        line for line in text.strip().splitlines() 
                        if line.strip() and not line.strip().startswith("#")
                    )
                    sections[section].append(f"{divider}\n{cleaned_lines}")
            elif section == "MITM":
                hostname_match = hostname_pattern.search(text)
                if hostname_match:
                    hostnames = hostname_match.group(1).replace("%APPEND%", "").split(',')
                    sections["MITM"].extend([hostname.strip() for hostname in hostnames if hostname.strip()])
        
        print(f"Successfully merged: {info['header']}")
        
    except requests.exceptions.RequestException as e:
        print(f"Failed to download {info['header']} file: {e}")

headers_combined = ", ".join(headers)

if sections["Rule"]:
    print(f"[Debug] Total Rule content to be saved: {len(sections['Rule'])} lines")
else:
    print("[Warning] No Rule content extracted")

os.makedirs('Chores/ruleset', exist_ok=True)
with open('Chores/ruleset/reject.list', 'w') as ruleset_file:
    ruleset_file.write("\n".join(sections["Rule"]))
print("Successfully saved [Rule] content to Chores/ruleset/reject.list")

unique_hostnames = list(dict.fromkeys(sections["MITM"]))
hostname_append_content = ", ".join(unique_hostnames)

current_date = datetime.now().strftime("%m/%d/%Y")

template_path = 'Chores/engineering/data/templates/All-in-One-2.x.sgmodule.template'
output_path = 'Surge/Module/All-in-One-2.x.sgmodule.sgmodule'

with open(template_path, 'r') as template_file:
    template_content = template_file.read()

for section, contents in sections.items():
    placeholder = f"{{{section}}}"
    section_content = "\n\n".join(contents) if section != "MITM" else contents
    template_content = template_content.replace(placeholder, str(section_content))

template_content = template_content.replace("{headers}", headers_combined)
template_content = template_content.replace("{hostname_append}", hostname_append_content)
template_content = template_content.replace("{{currentDate}}", current_date)

with open(output_path, 'w') as output_file:
    output_file.write(template_content)

print(f"File successfully merged and saved to: {output_path}")

File: Chores/engineering/merge.ts
import fs from 'fs';
import path from 'path';
import { promisify } from 'util';
import fetch from 'node-fetch';
import yaml from 'js-yaml';

interface SgmoduleInfo {
    url: string;
    header: string;
}

interface Sections {
    'URL Rewrite': string[];
    'Map Local': string[];
    'Script': string[];
    'MITM': string[];
    'Rule': string[];
}

const readFile = promisify(fs.readFile);
const writeFile = promisify(fs.writeFile);
const mkdir = promisify(fs.mkdir);

const MAX_DIVIDER_LENGTH = 30; // Maximum divider length (excluding header length)

async function mergeSgmodules() {
    try {
        // Read sgmodules.yaml
        const sgmoduleContent = await readFile('Chores/engineering/data/sgmodules.yaml', 'utf8');
        const sgmoduleInfo = yaml.load(sgmoduleContent) as SgmoduleInfo[];

        const sections: Sections = {
            'URL Rewrite': [],
            'Map Local': [],
            'Script': [],
            'MITM': [],
            'Rule': []
        };

        const headers: string[] = [];
        const sectionPattern = /\[(.*?)\]\s*\n(.*?)(?=\n\[|$)/gs;
        const hostnamePattern = /hostname\s*=\s*(.*)/i;

        // Process each URL in sgmodules.yaml
        for (const info of sgmoduleInfo) {
            try {
                const response = await fetch(info.url);
                if (!response.ok) {
                    throw new Error(`HTTP error! status: ${response.status}`);
                }
                const content = await response.text();
                console.log(`[Debug] Parsing content from ${info.header}`);

                headers.push(info.header);

                // Calculate dynamic divider
                const leftDashCount = Math.floor((MAX_DIVIDER_LENGTH - info.header.length) / 2);
                const rightDashCount = MAX_DIVIDER_LENGTH - info.header.length - leftDashCount;
                const divider = `# ${'-'.repeat(leftDashCount)} ${info.header} ${'-'.repeat(rightDashCount)}`;

                // Process each section
                let match;
                while ((match = sectionPattern.exec(content)) !== null) {
                    const [, section, text] = match;
                    if (section in sections && section !== 'MITM') {
                        if (section === 'Rule') {
                            // 清理注释和空行，只保留有效规则和模块头
                            const cleanedLines = text
                                .trim()
                                .split('\n')
                                .filter(line => {
                                    const trimmedLine = line.trim();
                                    // 保留非空行，且要么是模块头（以 # >> 开头）要么不是注释
                                    return trimmedLine && 
                                           (!trimmedLine.startsWith('#') || trimmedLine.startsWith('# >> '));
                                })
                                .join('\n');
                            // 添加模块头时使用正确的格式
                            sections.Rule.push(`# >> ${info.header}\n${cleanedLines}`);
                        }
                        else {
                            // Remove comment lines and empty lines, but keep divider
                            const cleanedLines = text
                                .trim()
                                .split('\n')
                                .filter(line => line.trim() && !line.trim().startsWith('#'))
                                .join('\n');
                            sections[section as keyof Sections].push(`${divider}\n${cleanedLines}`);
                        }
                    } else if (section === 'MITM') {
                        const hostnameMatch = text.match(hostnamePattern);
                        if (hostnameMatch) {
                            const hostnames = hostnameMatch[1]
                                .replace('%APPEND%', '')
                                .split(',')
                                .map(h => h.trim())
                                .filter(Boolean);
                            sections.MITM.push(...hostnames);
                        }
                    }
                }

                console.log(`Successfully merged: ${info.header}`);
            } catch (error) {
                console.error(`Failed to download ${info.header} file:`, error);
            }
        }

        const headersCombined = headers.join(', ');

        // Process and save rules
        if (sections.Rule.length > 0) {
        } else {
            console.log('[Warning] No Rule content extracted');
        }

        // Create ruleset directory if it doesn't exist
        await mkdir('Chores/ruleset', { recursive: true });

        // Process rules with comments and deduplication
        const processedRules = processRules(sections.Rule);
        await writeFile('Chores/ruleset/reject.list', processedRules.rejectRules);
        await writeFile('Chores/ruleset/direct.list', processedRules.directRules);
        await writeFile('Chores/ruleset/reject-drop.list', processedRules.rejectDropRules);
        await writeFile('Chores/ruleset/reject-tiny.list', processedRules.rejectTinyRules);
        await writeFile('Chores/ruleset/reject-no-drop.list', processedRules.rejectNoDropRules);

        // Process MITM hostnames
        const uniqueHostnames = Array.from(new Set(sections.MITM));
        const hostnameAppendContent = uniqueHostnames.join(', ');

        const currentDate = new Date().toLocaleDateString('en-US');

        // Read and process template
        const templateContent = await readFile(
            'Chores/engineering/templates/All-in-One-2.x.sgmodule.template',
            'utf8'
        );

        let outputContent = templateContent;
        for (const [section, contents] of Object.entries(sections)) {
            const placeholder = `{${section}}`;
            const sectionContent = section !== 'MITM' ? contents.join('\n\n') : contents;
            outputContent = outputContent.replace(placeholder, sectionContent);
        }

        outputContent = outputContent
            .replace('{headers}', headersCombined)
            .replace('{hostname_append}', hostnameAppendContent)
            .replace('{{currentDate}}', currentDate);

        await writeFile('Chores/sgmodule/All-in-One-2.x.sgmodule', outputContent);
        console.log('File successfully merged and saved to: Chores/sgmodule/All-in-One-2.x.sgmodule');
    } catch (error) {
        console.error('Error in mergeSgmodules:', error);
    }
}

interface ProcessedRules {
    rejectRules: string;
    directRules: string;
    rejectDropRules: string;
    rejectTinyRules: string;
    rejectNoDropRules: string;
}

function processRules(rules: string[]): ProcessedRules {
    const ruleMap = new Map<string, { rule: string }>();
    const rejectRules: string[] = [];
    const directRules: string[] = [];
    const rejectDropRules: string[] = [];
    const rejectTinyRules: string[] = [];
    const rejectNoDropRules: string[] = [];
    
    let currentModuleHeader: string | null = null;
    let isFirstModule = {
        reject: true,
        direct: true,
        rejectDrop: true,
        rejectTiny: true,
        rejectNoDrop: true
    };

    rules.forEach(ruleBlock => {
        const lines = ruleBlock.split('\n');
        currentModuleHeader = null;
        
        lines.forEach(line => {
            const trimmedLine = line.trim();
            if (!trimmedLine) return;

            if (trimmedLine.startsWith('# >> ')) {
                currentModuleHeader = trimmedLine;
                return;
            }

            if (trimmedLine && !trimmedLine.startsWith('#')) {
                // 移除策略组但保留参数
                const ruleParts = trimmedLine.split(',');
                const ruleType = ruleParts[0];
                const ruleValue = ruleParts[1];
                const originalPolicy = ruleParts[2];
                const ruleParams = ruleParts.slice(2).filter(param => 
                    ['no-resolve', 'extended-matching', 'pre-matching'].includes(param.toLowerCase())
                );
                
                // 重组规则，不包含策略组
                const newRule = [ruleType, ruleValue, ...ruleParams].join(',');
                const ruleWithoutPolicy = newRule.replace(/,(REJECT|DIRECT)(-[A-Z]+)?$/, '');

                // 根据不同策略组分类处理
                if (trimmedLine.includes('REJECT-DROP')) {
                    if (!ruleMap.has(ruleWithoutPolicy)) {
                        addRuleToList(rejectDropRules, currentModuleHeader, newRule, isFirstModule, 'rejectDrop');
                    }
                } else if (trimmedLine.includes('REJECT-TINY')) {
                    if (!ruleMap.has(ruleWithoutPolicy)) {
                        addRuleToList(rejectTinyRules, currentModuleHeader, newRule, isFirstModule, 'rejectTiny');
                    }
                } else if (trimmedLine.includes('REJECT-NO-DROP')) {
                    if (!ruleMap.has(ruleWithoutPolicy)) {
                        addRuleToList(rejectNoDropRules, currentModuleHeader, newRule, isFirstModule, 'rejectNoDrop');
                    }
                } else if (trimmedLine.includes('REJECT')) {
                    if (!ruleMap.has(ruleWithoutPolicy)) {
                        addRuleToList(rejectRules, currentModuleHeader, newRule, isFirstModule, 'reject');
                    }
                } else if (trimmedLine.includes('DIRECT')) {
                    if (!directRules.includes(newRule)) {
                        addRuleToList(directRules, currentModuleHeader, newRule, isFirstModule, 'direct');
                    }
                }

                if (!ruleMap.has(ruleWithoutPolicy)) {
                    ruleMap.set(ruleWithoutPolicy, { rule: ruleWithoutPolicy });
                }
            }
        });
    });

    return {
        rejectRules: rejectRules.join('\n'),
        directRules: directRules.join('\n'),
        rejectDropRules: rejectDropRules.join('\n'),
        rejectTinyRules: rejectTinyRules.join('\n'),
        rejectNoDropRules: rejectNoDropRules.join('\n')
    };
}

// 辅助函数：添加规则到对应列表
function addRuleToList(
    ruleList: string[], 
    header: string | null, 
    rule: string, 
    isFirstModule: {[key: string]: boolean}, 
    type: string
) {
    if (header && !ruleList.includes(header)) {
        if (!isFirstModule[type]) {
            ruleList.push('');
        }
        ruleList.push(header);
        isFirstModule[type] = false;
    }
    ruleList.push(rule);
}

// Execute the merge
mergeSgmodules().catch(console.error);
